### PR TITLE
realtek: reclaim completed TX buffers on RX poll

### PR DESCRIPTION
**Pull request by l0rdg3x**
### realtek: reclaim completed TX buffers on RX poll

Split out of #24400 per review. The conduit transmit path frees a TX skb only when its descriptor slot is reused, `RTETH_TX_RING_SIZE` transmissions later; until then the skb stays charged to the sending socket. Negligible with standard frames, but with jumbo MTUs a handful of completed transmissions exhausts the sender's budget: eight ~9k echo replies exceed the kernel ICMP socket limit (`2 * SKB_TRUESIZE(64 * 1024)`), after which every further reply of any size is dropped (`Icmp: OutErrors` grows, `OutMsgs` stalls) until unrelated transmissions recycle the slots. On the bench (RTL9303, Hasivo S1100W-8XGT-SE) this is a reproducible ICMP blackout after exactly 8 jumbo echo replies, healed deterministically by any 16 device-originated frames.

Design, addressing the review concerns on the first attempt:

- the driver-private ring info tracks the oldest pending slot and a pending count, so finding work costs one cached read per TX ring — while nothing is pending the poll takes no lock and never touches the uncached descriptor memory;
- the walk starts at the oldest pending slot and stops at the first descriptor the hardware still owns: uncached reads = completed slots + 1, not the whole ring;
- the TX queue lock is held only for that bounded walk;
- the transmit path reuses the same walk for the ring-wrap case instead of its old duplicated per-slot cleanup;
- stopped or watchdog-frozen queues are skipped, leaving `rteth_tx_timeout()` recovery untouched.

**Performance (RTL9303, iperf3 8 s × 3 runs, MTU 1500, median):**

| | master | with reclaim |
|---|---|---|
| TCP host→device | 170 Mbit/s | 173 Mbit/s (+1.8%) |
| TCP device→host | 115 Mbit/s | **128 Mbit/s (+11.3%)** |
| UDP 1400 device→host | 61.8 Mbit/s | 64.4 Mbit/s (+4.2%) |
| UDP 64 device→host | 3.01 Mbit/s | 3.18 Mbit/s (+5.6%) |
| ping RTT avg | 0.409 ms | 0.401 ms |

Device-sourced TCP improves because the socket's `sk_wmem` now drains as soon as a poll runs instead of up to 16 transmissions later, so the sender is throttled less. No path gets slower.

Functional check with the jumbo series from #24400 on top: the ICMP blackout is gone (40/40 sequential and 150/150 sustained ~9k echoes, was 8/40 and 2/150), byte-exact MTU enforcement unaffected, and jumbo iperf3 reaches 467/222 Mbit/s (host→device/device→host) at MTU 9000 versus 173/128 at 1500.

CC @plappermaul

*Bench testing and analysis assisted by Claude Fable 5 (Anthropic AI).*

### CODE REVIEW

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 1]
```diff
@@ -90,6 +90,8 @@ struct rteth_rx_info {
 
 struct rteth_tx_info {
 	int			slot;
+	int			dirty;   /* oldest slot with an unreleased skb */
+	unsigned int		pending; /* skbs not yet released */
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: slot/dirty/pending is partly redundant and currently need to cover edge cases. Use monotonic counters instead

- send_count: number of send/created SKBS/packets (instead of slot)
- clean_count: number of cleaned/reclaimed SKBS

With this: 

- Just increase counters with "++" (no wrap needed)
- int slot = tx_info->clean_count & (RTETH_TX_RING_SIZE - 1); (or send_coount)
- pending = tx_info->send_count - tx_info->clean_count;

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 2]
```diff
@@ -1107,6 +1162,9 @@ static int rteth_start_xmit(struct sk_buff *skb, struct net_device *dev)
 	dma_wmb();
 	ctrl->tx_data[ring].ring[slot] = packet_dma | RING_OWN_HW;
 	ctrl->tx_info[ring].slot = (slot + 1) % RTETH_TX_RING_SIZE;
+	if (!ctrl->tx_info[ring].pending)
+		ctrl->tx_info[ring].dirty = slot;
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: With above idea unneeded edge case.

**Review by plappermaul — CHANGES_REQUESTED** [review]
Much better implementation. Some nits left. 

Please attach a performance test for the new version too.

**Comment by l0rdg3x** [comment]
Done — `slot`/`dirty`/`pending` replaced by monotonic `send_count`/`clean_count` (difference = outstanding skbs, low bits = ring slot, plus a `BUILD_BUG_ON` guarding the power-of-two assumption). The initialization edge case is gone, the transmit path is a single `send_count++`.

Performance re-run for this version, same protocol (iperf3 8 s × 3 runs, MTU 1500, two boots, median):

| | master | monotonic-counter reclaim |
|---|---|---|
| TCP host→device | 170 Mbit/s | 171 Mbit/s |
| TCP device→host | 115 Mbit/s | 119 Mbit/s (+3.5%) |
| UDP 1400 device→host | 61.8 Mbit/s | 64.0 Mbit/s (+3.6%) |
| UDP 64 device→host | 3.01 Mbit/s | 3.16 Mbit/s (+5%) |
| ping RTT avg | 0.409 ms | 0.396 ms |

No path regresses. For the record, the previous revision measured a couple of percent higher on device-sourced TCP (128 vs 119) across two boots each — the logic is equivalent and lighter now, so this smells like structure-layout/alignment noise rather than anything algorithmic; I kept your version since it is the cleaner one.

Functional check with the #24400 series on top unchanged: 40/40 sequential and 150/150 sustained ~9k echoes, byte-exact enforcement intact.

*Bench testing and analysis assisted by Claude Fable 5 (Anthropic AI).*

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 3]
```diff
@@ -657,6 +658,59 @@ static void rteth_free_tx_buffers(struct rteth_ctrl *ctrl)
...
+
+		dma_unmap_single(&ctrl->pdev->dev, ctrl->tx_data[r].frag[i].dma,
+				 tx_info->skb[i]->len, DMA_TO_DEVICE);
+		dev_consume_skb_any(tx_info->skb[i]);
+		tx_info->skb[i] = NULL;
+		tx_info->clean_count++;
+	}
+}
+
+static void rteth_reclaim_tx_buffers(struct rteth_ctrl *ctrl)
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: rename to rteth_reclaim_tx_rings() to align with above function.

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 4]
```diff
@@ -657,6 +658,59 @@ static void rteth_free_tx_buffers(struct rteth_ctrl *ctrl)
...
+	}
+}
+
+/* Release the contiguous run of transmitted skbs. Until released they stay
+ * charged to the sending socket, and with jumbo MTUs a handful of completed
+ * transmissions can exhaust the sender's budget, e.g. eight ~9k echo replies
+ * exceed the kernel ICMP socket limit and every further reply is dropped.
+ * The walk starts at the oldest unreleased slot and stops at the first
+ * descriptor the hardware still owns. The caller holds the TX queue lock.
+ */
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: This function is self explanatory. Drop this comment.

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 5]
```diff
@@ -657,6 +658,59 @@ static void rteth_free_tx_buffers(struct rteth_ctrl *ctrl)
...
+		    READ_ONCE(ctrl->tx_info[r].clean_count))
+			continue;
+
+		txq = netdev_get_tx_queue(ctrl->dev, r);
+
+		__netif_tx_lock(txq, smp_processor_id());
+
+		/* Leave stopped or watchdog-frozen queues to the recovery
+		 * path: rteth_tx_timeout() releases every buffer itself.
+		 */
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: Drop comment. Code tells the same.

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 6]
```diff
@@ -1086,11 +1141,11 @@ static int rteth_start_xmit(struct sk_buff *skb, struct net_device *dev)
...
 	}
 
-	if (likely(*packet_skb)) {
-		/* cleanup old data of this slot */
-		dma_unmap_single(&ctrl->pdev->dev, frag->dma, (*packet_skb)->len, DMA_TO_DEVICE);
-		dev_consume_skb_any(*packet_skb);
-	}
+	/* Ring wrapped before any reclaim ran: this slot's descriptor is no
+	 * longer hardware-owned, so the walk releases it (and any older run).
+	 */
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: Drop comment. Code below stands for itself.

File: target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c [hunk 7]
```diff
@@ -1297,6 +1352,11 @@ static int rteth_poll_rx(struct napi_struct *napi, int budget)
 	struct rteth_ctrl *ctrl = rx_q->ctrl;
 	int work_done, ring = rx_q->id;
 
+	/* Ahead of the receive pass, so replies generated while handling it
+	 * always find their socket send budget released.
+	 */
```
Comment (plappermaul) [inline target/linux/realtek/files-6.18/drivers/net/ethernet/rtl838x_eth.c]: Drop comment.

**Review by plappermaul — CHANGES_REQUESTED** [review]
One round of final cleanups and we are ready to go.

**Comment by l0rdg3x** [comment]
Done, all five addressed.

**Comment by plappermaul** [comment]
Thanks! Rebased on top of main and merged!

**Comment by l0rdg3x** [comment]
You are welcome! :)
